ENH: Show unit information in repr for datetime64("NaT")

riku-sakamoto opened the pull request:
## Overview

* Fix #28496 

## Details

* Unlike timedelta objects, datetime64 does not display its unit in the repr output (see #28496). 
* To maintain backward compatibility, this change adds unit information only when the datetime64 object has a NaT value.

riku-sakamoto committed:
ENH: Show unit information in repr for datetime64("NaT")

Unlike timedelta objects, datetime64 does not display its unit in the repr output (see #28496).
To maintain backward compatibility, this change adds unit information only when the datetime64 object has a NaT value.

ngoldbaum reviewed (commented):
Makes sense to me. I think because this changes the repr it needs a release note, see `doc/release/upcoming_changes/README.rst` for more info about adding release note blurbs to PRs.

ngoldbaum commented on numpy/_core/src/multiarray/scalartypes.c.src:
```diff
@@ -939,12 +939,23 @@ datetimetype_repr(PyObject *self)
+
+        PyObject *meta = metastr_to_unicode(&scal->obmeta, 1);
+        if((scal->obval == NPY_DATETIME_NAT) && (meta != NULL)){
+            if (legacy_print_mode > 125) {
```
I guess technically we could define a new numpy 2.3 legacy print mode as well. I don't know if it's worth the ceremony just for this.

riku-sakamoto committed:
DOC: add release note for showing unit in `__repr__` of datatime64("NaT")

Add a new release note for the change that shows time unit in the __repr__
output of datetime64("NaT"). See gh-29396.

ngoldbaum commented:
Let's bring it in. If anyone reads this in the future because of downstream test failures please open an issue and we can reconsider adding a new legacy print mode.

ngoldbaum commented:
Thanks @riku-sakamoto
